elastic/kibana
# [Security Solution] Fixes a bug with timeline sourcerer state

## Summary

When trying to open a saved timeline with a sourcerer set to NOT default data view id and a narrowed down index pattern, I noticed the timeline's index patterns were getting reset to the data view's entire pattern list. There were 3 problems that contributed:

**Problem 1**: When checking that we have initialized each active data view, we call initialize if the timeline scope pattern list is empty. However, if a saved timeline is being initialized it  the timeline scope will already pattern list before the data view is initialized
**Solution**: Instead of checking if the timeline scope pattern list is empty, check if the timeline data view pattern list is empty.

**Problem 2**: After initializing the data view, we update the scope. However, if the scope was set by a saved timeline the pattern list gets overwritten
**Solution**: Add a check that if the timeline scope data view id is the same as the data view id we are going to initialize and the timeline scope has selected patterns, do not update the scope after initializing the data view. 

**Problem 3**: When scope selected patterns are set by the saved timeline, we attempt to validate the selected patterns against the data view pattern list. However, if the data view has not been initialized this comparison is invalid
**Solution**: Add a check for data view pattern list length before comparing against selected patterns 

### Code cleanup
The sourcerer actions `updateSourcererDataViews` and `setDataView` were doing the same thing, so I consolidated.

### How to reproduce bug on `main`

1. Create a data view with multiple active index patterns
2. go to the timelines page (refresh after creating data view)
3. open timeline, save it
4. Set timeline sourcerer to your new data view and narrow the index pattern down to a single index
5. Refresh the page
6. Open the timeline (it should be already active in timeline bottom bar)
7. **Bug**: Timeline sourcerer data view is correct but index pattern has been reset

### How to test

1. Create a data view with multiple active index patterns
2. go to the timelines page (refresh after creating data view)
3. open timeline, save it
4. Set timeline sourcerer to your new data view and narrow the index pattern down to a single index
5. Refresh the page
6. Open the timeline (it should be already active in timeline bottom bar)
7. **Fix**: Ensure that the timeline sourcerer data view is correct and the index pattern is narrowed down to the single index


#### Before
![page_refresh](https://user-images.githubusercontent.com/6935300/195439216-4d0f7e65-655c-485a-945c-cedda572159f.gif)

#### After
![after](https://user-images.githubusercontent.com/6935300/195439208-9df029b0-556b-4542-9bd1-92e430682e85.gif)
